kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -350,59 +385,115 @@ internal suspend fun <T: AbstractCoroutine<Unit>> CoroutineScope.runTestCoroutin
+        /**
+         * The thread in which the task was last seen executing.
+         */
+        val lastKnownPosition = MutableStateFlow<Any?>(null)
+        /**
+         * We run the tasks in the test coroutine using [Dispatchers.Default]. On JS, this does nothing particularly,
+         * but on the JVM and Native, this means that the timeout can be processed even while the test runner is busy
+         * doing some synchronous work.
+         */
+        val workRunner = launch(Dispatchers.Default + CoroutineName("kotlinx.coroutines.test runner")) {
             while (true) {
-                scheduler.tryRunNextTaskUnless { !isActive }
-                // yield so that the `select` below has a chance to check if its conditions are fulfilled
-                yield()
+                lastKnownPosition.value = getLastKnownPosition()
+                val executedSomething = scheduler.tryRunNextTaskUnless { !isActive }
+                if (executedSomething) {
+                    // yield to check for cancellation
+                    yield()
+                } else {
+                    // no more tasks, we should suspend until there are some more
+                    scheduler.receiveDispatchEvent()
+                }
             }
         }
         try {
-            select<Unit> {
-                coroutine.onJoin {
-                    // observe that someone completed the test coroutine and leave without waiting for the timeout
-                    completed = true
-                }
-                scheduler.onDispatchEvent {
-                    // we received knowledge that `scheduler` observed a dispatch event, so we reset the timeout
-                }
-                onTimeout(dispatchTimeout) {
-                    handleTimeout(coroutine, dispatchTimeout, tryGetCompletionCause, cleanup)
-                }
+            withTimeout(testTimeout) {
+                coroutine.join()
+                workRunner.cancelAndJoin()
             }
-        } finally {
-            backgroundWorkRunner.cancelAndJoin()
+        } catch (_: TimeoutCancellationException) {
+            val exception = handleTimeout(coroutine, testTimeout, tryGetCompletionCause, cleanup)
+            dumpCoroutines()
+            /**
+             * There's a race that may lead to the misleading results here, but it's better than nothing.
+             * The race: `lastKnownPosition` is read, then the task executed in `workRunner` completes,
+             * then `updateStacktrace` does its thing, but the thread is already busy doing something else.
+             */
+            updateStacktrace(exception, lastKnownPosition.value)
+            // we can't abandon the work we're doing, so if it hanged, we'll still hang, despite the timeout.
+            (coroutine as CoroutineScope).cancel("The test timed out", exception)
+            coroutine.join()
+            workRunner.cancelAndJoin()
+            scheduler.advanceUntilIdleOr { false }
```

Comment:
Consider the following scenario: the coroutine is timed out and properly joined, but some auxiliary watchdog wasn't cancelled. Does it mean that here we'll hang? 

I do not object against this behaviour, rather pointing out potential scenarios

Reply:
If something's executing in the test scheduler and doesn't respond to cancellation, yes, we'll keep executing it even if it means that we'll hang.